File: inlay/android/src/main/kotlin/co/leancode/inlay/FlutterRoute.kt
package co.leancode.inlay

import co.leancode.inlay.navigator.PageSettings

interface FlutterRoute {
    fun toPageSettings(): PageSettings
}

interface FlutterDialogRoute {
    fun toPageSettings(): PageSettings
}

/**
 * A [FlutterRoute] whose screen returns a typed result of type [R].
 *
 * Generated route classes with a `result:` type implement this interface,
 * which lets [InlayNavigator.push] and the embedding wrappers deliver an
 * already-decoded `R?` instead of a raw `Any?`.
 */
interface FlutterRouteWithResult<R : Any> : FlutterRoute {
    /** Decodes the wire-format result into [R]; `null` stays `null`. */
    fun decodeResult(raw: Any?): R?
}

/**
 * A [FlutterDialogRoute] whose dialog returns a typed result of type [R].
 */
interface FlutterDialogRouteWithResult<R : Any> : FlutterDialogRoute {
    /** Decodes the wire-format result into [R]; `null` stays `null`. */
    fun decodeResult(raw: Any?): R?
}

File: inlay/ios/inlay/Sources/inlay/FlutterRoute.swift
import Foundation

public protocol FlutterRoute {
    func toPageSettings() -> PageSettings
}

public protocol FlutterDialogRoute {
    func toPageSettings() -> PageSettings
}

/// A ``FlutterRoute`` whose screen returns a typed result.
///
/// Generated route structs with a `result:` type conform to this protocol
/// (`ResultValue` is inferred from the generated `decodeResult`), which lets
/// `InlayNavigator.push` and the SwiftUI wrappers deliver an already-decoded
/// `ResultValue?` instead of a raw `Any?`.
public protocol FlutterRouteWithResult: FlutterRoute {
    associatedtype ResultValue
    /// Decodes the wire-format result into `ResultValue`; `nil` stays `nil`.
    static func decodeResult(_ raw: Any?) -> ResultValue?
}

/// A ``FlutterDialogRoute`` whose dialog returns a typed result.
public protocol FlutterDialogRouteWithResult: FlutterDialogRoute {
    associatedtype ResultValue
    /// Decodes the wire-format result into `ResultValue`; `nil` stays `nil`.
    static func decodeResult(_ raw: Any?) -> ResultValue?
}
